src/components/views/rooms/DecryptionFailureBar.tsx

```diff
@@ -0,0 +1,208 @@
+/*
+Copyright 2022 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import React, { useCallback, useContext, useEffect, useState } from 'react';
+import { Room } from 'matrix-js-sdk/src/models/room';
+import { MatrixEvent } from 'matrix-js-sdk/src/models/event';
+
+import Modal from '../../../Modal';
+import { _t } from '../../../languageHandler';
+import defaultDispatcher from "../../../dispatcher/dispatcher";
+import { Action } from "../../../dispatcher/actions";
+import AccessibleButton from '../elements/AccessibleButton';
+import { OpenToTabPayload } from "../../../dispatcher/payloads/OpenToTabPayload";
+import { UserTab } from "../dialogs/UserTab";
+import MatrixClientContext from "../../../contexts/MatrixClientContext";
+import SetupEncryptionDialog from '../dialogs/security/SetupEncryptionDialog';
+import { SetupEncryptionStore } from '../../../stores/SetupEncryptionStore';
+import Spinner from "../elements/Spinner";
+
+interface IProps {
+    room: Room;
+    failures: MatrixEvent[];
+}
+
+// Number of milliseconds to display a loading spinner before prompting the user for action
+const WAIT_PERIOD = 5000;
+
+export const DecryptionFailureBar: React.FC<IProps> = ({ failures, room }) => {
+    const context = useContext(MatrixClientContext);
+
+    const [waiting, setWaiting] = useState<boolean>(true);
+    useEffect(() => {
+        const timeout = setTimeout(() => setWaiting(false), WAIT_PERIOD);
+        return () => clearTimeout(timeout);
+    }, []);
+
+    const [needsVerification, setNeedsVerification] = useState<boolean>(false);
+    const [hasOtherVerifiedDevices, setHasOtherVerifiedDevices] = useState<boolean>(false);
+    const [hasKeyBackup, setHasKeyBackup] = useState<boolean>(false);
+
+    const [requestedSessions, setRequestedSessions] = useState<Set<string>>(new Set());
+
+    useEffect(() => {
+        if (needsVerification || !hasOtherVerifiedDevices) return;
+
+        const newRequestedSessions = new Set<string>();
+        for (const event of failures) {
+            const sessionId = event.getWireContent().session_id;
+            if (!sessionId || requestedSessions.has(sessionId) || newRequestedSessions.has(sessionId)) continue;
+            newRequestedSessions.add(sessionId);
+            context.cancelAndResendEventRoomKeyRequest(event);
```

Comment:
Slightly struggling to follow what's going on here (more comments, please), but doesn't this mean that we'll cancel and resend room key requests every time a decryption error comes into view?

Generally: it feels extremely surprising, architecturally speaking, that the failure bar has responsibility for sending out the room key requests. I would expect it just to be a view without side-effects, and managing the room key requests to happen elsewhere.

Reply:
This is replacing the individual "click here to resend requests" links on every single undecryptable message, with the observations that
- This should be per-session, not per-message
- Rather than waiting for the user to tell us to do the obviously helpful thing, we should just do it automatically.
This component is aggregating information on all the decryption failures, as well as other important information like whether we're even verified and whether there's anyone to send requests to.

This means that we'll cancel and resend room key requests every time the user scrolls back to a decryption error they've scrolled away from. If there were already other decryption errors on screen from the same session, then we won't cancel or resend anything.

Reply:
> Rather than waiting for the user to tell us to do the obviously helpful thing, we should just do it automatically.

I don't think I agree with this, at all.

We're already automatically sending out room key requests when we get a decryption failure. I don't think that's a good idea in the first place (first, it causes a huge amount of inter-device traffic, which ends up making decryption failures *more* likely because it takes ages to transfer and handle all those to-device messages; second, it ends up papering over actual bugs in the E2EE implementation, so those bugs happen intermittently but are harder to track down and fix).

By repeating the action every single time a decryption error comes into view, we're making the whole thing worse. I'd really like us to discuss this as a team before proceeding.

Reply:
We discussed this in some detail in chat. A summary is something like:
 * I was concerned that this change would add significantly to the number of room-key-request messages we are sending (which is already dangerously high).
 * @duxovni suggested she could add a button to the Decryption Failure bar instead of sending the requests automatically
 * That sounded good to me, but I requested that the button be hidden if all of the UTD events already have outstanding requests.

Reply:
Done.